File: Minimize Digit Sum/MINDIGSUM.java
//minimize digit sum; codechef : 07.09.21; complete : n

import java.util.*;

class Main {
	public static void main(String args[]) {
		int q, n, l, r;
		Scanner S = new Scanner(System.in);

		q = 0;
		if(S.hasNextInt()) q = S.nextInt();

		while(q != 0) {
			n = 0;
			if(S.hasNextInt()) n = S.nextInt();

			l = 0;
			if(S.hasNextInt()) l = S.nextInt();

			r = 0;
			if(S.hasNextInt()) r = S.nextInt();

			System.out.println(MINDIDSUM(n, l, r));

			q -= 1;
		}
	}

	public static int MINDIDSUM(int n, int l, int r) {
		int k = 0, MINBASE = 0, MAX = Integer.MAX_VALUE;

		if(l <= n && r>= n) return n;
		if(l >= n) return l;

		for(int cBase = r; cBase >=l; cBase--) {

			if(r >= n && n % cBase == 0) return cBase;
			
			k = baseConversion(n, cBase, MAX);
			// System.out.println("digit sum : "+k);
			if(k < MAX) {
				MAX = k;
				MINBASE = cBase;
			}

			if(MAX == 1) break;
		}
		return MINBASE;
	}


	public static int baseConversion(int n, int Base, int MAX) {
		int digitSum = 0, number = n;

		while(number != 0) {
			digitSum += number % Base;
			number /= Base;

			if(digitSum >= MAX) break;
		}

		return digitSum;
	}
}

File: Minimize Digit Sum/MINDIGSUM.c
// minimize digit sum; codechef : 07.09.21; complete : y

#include<stdio.h>
#include<limits.h>
int main(void)
{
	int mindigsum(int , int , int );
	int q, n, l, r;
	scanf("%d", &q);

	while(q != 0) {
		scanf("%d %d %d", &n, &l, &r);
		printf("%d\n", mindigsum(n, l, r));
		q -= 1;
	}
	return 0;
}

int mindigsum(int n, int l, int r) {
	int min, sum, sum2, number;

	if(l <= n && r>= n) return n;
	if(l > n) return l;

	sum2 = INT_MAX;
	min = 0;

	while(l < r && n/r < r) {

		int lvalue = n/r, rvalue = n % r;

		if(sum2 > lvalue + rvalue) {
			sum2 = lvalue + rvalue;
			min = r;
		}
		r = n/(lvalue+1);
	}

	while(l <= r) {
		number = n;
		sum = 0;

		while(1) {
			if(number < l) {
				sum += number;

				if(sum < sum2) {
					min = l;
					sum2 = sum;
				}
				break;
			}

			sum += number % l;
			number /= l;
			if(sum >= sum2) break;
		}

		l += 1;
	}

	return min;
}